src/core/symbol/MerkleHashBuilder.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 SYMBOL
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { sha3_256 } from 'js-sha3';
+
+export class MerkleHashBuilder {
```

Comment:
In general, I would prefer functional stateless methods that receive a list of Uint8Array and returns a Uint8Array instead of stateful mutable builder objects. 

Example:

```
 
    public static calculateRootHash(hashes: Uint8Array[]): Uint8Array {
        if (hashes.length === 0) {
            return new Uint8Array(32);
        }
        let numRemainingHashes = hashes.length;
        while (numRemainingHashes > 1) {
            for (let i = 0; i < numRemainingHashes; i += 2) {
                if (i + 1 < numRemainingHashes) {
                    hashes.splice(i / 2, 0, Converter.hash(hashes[i], hashes[i + 1]));
                } else {
                    hashes.splice(i / 2, 0, Converter.hash(hashes[i], hashes[i]));
                    ++numRemainingHashes;
                }
            }
            numRemainingHashes /= 2;
        }
        return hashes[0];
    }
    ```

Something similar with the hashing operations.

Reply:
I don't see much difference tbh,  you can do builder.update().update().final(). Otherwise, the hash array has to be built before calling. perhaps this is the main difference?

Reply:
It's programming style. Stateless and immutable code has better caracteristics in my opinion. It is just inputs and outputs, you don't need to keep updating, resetting, and finalizing an object. You depend on the object's current state in order to know what's going to do.

Reply:
I am okay with either tbh. it is builder.update ve array.push, what do you think @Jaguar0625 ?

Reply:
One similar example extracted from 2 implementations of a hasher. One is a builder style object, the other one is just a method. Old SDK, new SDK PR
```
 const entityHashBytes: Uint8Array = new Uint8Array(
      signature.length + publicKey.length + generationHash.length + transactionBody.length,
  );
  entityHashBytes.set(signature, 0);
  entityHashBytes.set(publicKey, pubKeyIdx);
  entityHashBytes.set(generationHash, generationHashIdx);
  entityHashBytes.set(transactionBody, transactionBodyIdx);

  // 6) create SHA3 hash of transaction data
  // Note: Transaction hashing always uses SHA3
  SHA3Hasher.func(entityHash, entityHashBytes, 32);
  return entityHash
```


 VS
```
return Converter.hash(signature, publicKey, generationHash, transactionBody);
```

Reply:
> I am okay with either tbh. it is builder.update ve array.push, what do you think @Jaguar0625 ?

There isn't array push, it's mapping one object to another while keeping the immutability.

Like:

https://github.com/symbol/sdk-typescript/pull/25/files#diff-269e0a793317b6143e6114669f2222748ab7016cdc57be29b566fb21e349e625R156

Reply:
> In general, I would prefer functional stateless methods that receive a list of Uint8Array and returns a Uint8Array instead of stateful mutable builder objects.

stateless would imply that you do not overwrite hashes and do calculations recursively, the difference here would be memory usage O(n) in "stateful" vs O(n log n) in "stateless"
also talking about stateless in a method that simply does all calculations within it's call is at least disputable...

as for push() on interface level vs passing all hashes - you'd still need to collect those hashes _somewhere_ (don't mind)

Reply:
Hi @gimer , thank you for your comment.

What I recommended was to have a stateless client API/Contract rather than a mutable object/builder (API level, not implementation level).

Something like:

````
// imutable api.
public static calculateRootHash(hashes: Uint8Array[]): Uint8Array {
 /// may use local mutable state, same implementation than final() mthod just that hashes is a param.
}
````
Rather than:
````

// mutable object, client needs to maintain the state
export class MerkleHashBuilder {

    public final(): Uint8Array {
    ....
    }

    public update(hash: Uint8Array): MerkleHashBuilder {
     ....   
    }

}

The implementation could have mutable local fields for performance reasons, the same implementation that the builder, just that the user wouldn't need to know about the builder's state.

I do agree that going full recursive/Haskell style implementation has a performance impact as described. The recommendation was just an API recommendation.